keps/sig-network/20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -0,0 +1,631 @@
+```
+  --bind-address  stringSlice   (IP addresses, in a comma separated list, Default: [0.0.0.0,])
+```
+Only the first address of each IP family will be used; all others will be ignored.
+
+#### Multiple cluster CIDRs configuration
+
+The existing "--cluster-cidr" option for the [cloud-controller-manager](https://kubernetes.io/docs/reference/command-line-tools-reference/cloud-controller-manager/) will be modified to support multiple IP CIDRs in a comma-separated list (rather than a single IP CIDR).
+```
+  --cluster-cidr  ipNetSlice   (IP CIDRs, in a comma separated list, Default: [])
+```
+Only the first CIDR for each IP family will be used; all others will be ignored.
+
+The cloud_cidr_allocator will be updated to support allocating from multiple CIDRs. The route_controller will be updated to create routes for multiple CIDRs.
+
+### Container Environment Variables
+
+The [container environmental variables](https://kubernetes.io/docs/concepts/containers/container-environment-variables/#container-environment) should support dual stack.
+
+Pod information is exposed through environmental variables on the pod. There are a few environmental variables that are automatically created, and some need to be specified in the pod definition, through the downward api.
+
+The Downward API [status.podIP](https://kubernetes.io/docs/tasks/inject-data-application/downward-api-volume-expose-pod-information/#capabilities-of-the-downward-api) will preserve the existing single IP address, and will be set to the default IP for each pod. A new environmental variable named status.podIPs will contain a comma-separated list of IP addresses. The new pod API will have a slice of structures for the additional IP addresses. Kubelet will translate the pod structures and return podIPs as a comma-delimited string.
+
+Here is an example of how to define a pluralized MY_POD_IPS environmental variable in a pod definition yaml file:
+```
+  - name: MY_POD_IPS
+    valueFrom:
+      fieldRef:
+        fieldPath: status.podIPs
+  - name: MY_POD_IPS_PROPERTIES_0
+    valueFrom:
+      fieldRef:
+        fieldPath: status.podIPs[0].properties
+```
+
+This definition will cause an environmental variable setting in the pod similar to the following:
+```
+MY_POD_IPS=fd00:10:20:0:3::3,10.20.3.3
+MY_POD_IPS_PROPERTIES_<IP_index>=x:y #We assume that colon is not going to used in map values.
+```
+
+### Kubeadm Support
+
+Dual-stack support will need to be added to kubeadm both for dual-stack development purposes, and for use in dual-stack continuous integration tests.
+
+- The Kubeadm config options and config file will support dual stack options for apiserver-advertise-address, and podSubnet.
+
+#### Kubeadm Configuration Options
+
+The kubeadm configuration options for advertiseAddress and podSubnet will need to be changed to handle a comma-separated list of CIDRs:
+```
+    api:
+      advertiseAddress: "fd00:90::2,10.90.0.2" [Multiple IP CIDRs, comma separated list of CIDRs]
```

Comment:
currently in `kubeadm` [if the `advertiseAddress` is empty it takes the first IPv4 address ](https://github.com/kubernetes-sigs/kind/pull/348#discussion_r260159854)
We have to change this behavior to support this, and this makes me think about all the defaults option of the different components (kubelet, ...) that IMHO is is too big to define in this KEP. 

Should we add a section to define what will be the configuration options default behavior?, i.e.. 

```
### Configuration options defaults

For all the configuration options that require an IP address or Subnet, in case this option is not provided or set, it should look equally for both IPv4 and IPv6 addresses or networks.
```

Reply:
> [Looking at the code, we are giving priority to IPv4 over IPv6](https://github.com/kubernetes/kubernetes/blob/a3ccea9d8743f2ff82e41b6c2af6dc2c41dc7b10/staging/src/k8s.io/apimachinery/pkg/util/net/interface.go#L321) , do we need to change this behavior?

no, we cannot. ipv4 should still take precedence over ipv6. 
ipv6 needs to be explicit.

from the KEP:
> The kubeadm configuration options for advertiseAddress and podSubnet will need to be changed to handle a comma-separated list of CIDRs:

ideally, this has to be discussed during a kubeadm office hours meeting before jumping into PRs.
https://docs.google.com/document/d/130_kiXjG7graFNSnIAgtMS1G8zPDwpkshgfRYS0nggo/edit

overall dual stack is out-of-scope for the kubeadm team for this cycle, because we are overbooked.

Reply:
Thanks @neolit123. The current plan is that once we have support for dual-stack in core Kubernetes with the appropriate documentation we will focus on helping the tooling ecosystem get onboard. In the case of kubeadm this might be a little different if it's absolutely necessary for the purposes of testing. We will be sure to show up at the office hours and determine the best course of action in close consult with the team that works on kubeadm.